google-cloud-firestore/src/main/java/com/google/cloud/firestore/CollectionGroup.java

```diff
@@ -102,21 +104,32 @@ public ApiFuture<List<QueryPartition>> getPartitions(long desiredPartitionCount)
     } else {
       PartitionQueryRequest request = buildRequest(desiredPartitionCount);
 
-      try {
-        return ApiFutures.transform(
-            rpcContext.sendRequest(request, rpcContext.getClient().partitionQueryPagedCallable()),
-            response -> {
-              final ImmutableList.Builder<QueryPartition> partitions = ImmutableList.builder();
-              consumePartitions(
-                  response,
-                  queryPartition -> {
-                    partitions.add(queryPartition);
-                    return null;
-                  });
-              return partitions.build();
-            },
-            MoreExecutors.directExecutor());
+      TraceUtil.Span span =
+          rpcContext
+              .getFirestore()
+              .getOptions()
+              .getTraceUtil()
+              .startSpan(TraceUtil.SPAN_NAME_PARTITION_QUERY);
+      try (Scope ignored = span.makeCurrent()) {
+        ApiFuture<List<QueryPartition>> result =
+            ApiFutures.transform(
+                rpcContext.sendRequest(
+                    request, rpcContext.getClient().partitionQueryPagedCallable()),
+                response -> {
+                  final ImmutableList.Builder<QueryPartition> partitions = ImmutableList.builder();
+                  consumePartitions(
+                      response,
+                      queryPartition -> {
+                        partitions.add(queryPartition);
+                        return null;
+                      });
+                  return partitions.build();
+                },
+                MoreExecutors.directExecutor());
+        span.endAtFuture(result);
+        return result;
       } catch (ApiException exception) {
+        span.end(exception);
```

Comment:
In openCensus, we do `span.end()` in the `finally` block:https://github.com/googleapis/java-firestore/pull/1589/files#diff-04b7fd9ffd834a763fe56dd88c3e16d42bd8d1d3e3df4b134ae9ad296f147c00L133

Will there be an edge case where exception thrown but not `ApiException` type, and we end up failing to end the span?

Reply:
yeah, I think that's a legitimate corner case. There's a subtle difference with OpenCensus -- which is that with open-telemetry we end the span differently based on whether there was an exception or not. So I will add a `catch` block to catch all other kinds of Throwables, instead of adding a `finally` block.